sound/soc/sof/control.c

```diff
@@ -51,29 +51,40 @@ int snd_sof_volume_get(struct snd_kcontrol *kcontrol,
 	unsigned int i, channels = scontrol->num_channels;
 	int err, ret;
 
-	ret = pm_runtime_get_sync(sdev->dev);
-	if (ret < 0) {
-		dev_err(sdev->dev, "error: volume get failed to resume %d\n",
-			ret);
-		return ret;
+	/* if dsp control data is not cached, wake it up and get the data */
+	if (!scontrol->is_dsp_resident) {
```

Comment:
@bardliao @lgirdwood should we always use cached data or only when the device is runtime suspended?

Reply:
@bardliao I also think this check for is_dsp_resident is redundant. the kcontrol data is always cached including at first boot when alsa restores all kcontrols to the previously stored values. 

So in my opinion, we should check if the device is suspended and if it is, we set the cached values and if not we should get the value from the dsp. What do you think?

Reply:
@ranj063  I think we can always use cached data. If it is not a volatile kcontrol, the cached data should be always aligned with dsp. On the other hand, we should always get the data from dsp if it is a volatile kcontrol.  Regarding the is_dsp_resident flag, what will happen after factory reset? We may not set all kcontrols in the UCM (or similar) file. Also, I think Android will not restore "ALL" kcontrols. In my opinion, driver should guarantee the cached data is aligned with dsp. I think alternative solution is to set the default value to both dsp and cache when topology is loaded. I.e. The initial value will be decided by topology instead of dsp firmware.

Reply:
@bardliao @ranj063 cache should always be used for readback unless kcontrol is volatile, same rules as regmap. writes updaes cache and FW iff component is active.